src/conventional/amdar_bufr2ioda.py

```diff
@@ -0,0 +1,505 @@
+            vals[variable] = np.full(len(year), float_missing_value, dtype=np.float32)
+
+    # Be done with this BUFR message.
+    codes_release(bufr)
+
+    # For any of the MetaData elements that were totally lacking, fill entire vectory with missing.
+    for k, v in metaDataKeyList.items():
+        if not any(meta_data[k]):
+            meta_data[k] = assign_missing_meta(k, year)
+
+    # Need to transform some variables to others (wind speed/direction to components for example).
+    uwnd = np.full(len(year), float_missing_value)
+    vwnd = np.full(len(year), float_missing_value)
+    n = 0
+    for wind_direction in vals['windDirection']:
+        if (wind_direction != float_missing_value and vals['windSpeed'][n] != float_missing_value):
+            uwnd[n], vwnd[n] = met_utils.dir_speed_2_uv(wind_direction, vals['windSpeed'][n])
+        n+=1
+
+    spfh = np.full(len(year), float_missing_value)
+    n = 0
+    for mixing_ratio in vals['mixingRatio']:
+        if (mixing_ratio>0 and mixing_ratio<25.E-3):
+            spfh[n] = mixing_ratio / (1.0 + mixing_ratio)
+        n+=1
+
+    # Move everything into the final data dictionary, including metadata.
+    data['eastward_wind'] = np.append(data['eastward_wind'], uwnd)
+    data['northward_wind'] = np.append(data['northward_wind'], vwnd)
+    data['specific_humidity'] = np.append(data['specific_humidity'], spfh)
+    data['air_temperature'] = np.append(data['air_temperature'], vals['airTemperature'])
+    for key in meta_keys:
+        data[key] = np.append(data[key], meta_data[key])
+
+    if debug: print ("BUFR message number: ", count[0])
+    if debug: print ("number of observations so far: ", count[1])
+    if debug: print ("number of invalid or useless observations: ", count[2])
+    return data, count, start_pos
+
+
+if __name__ == "__main__":
+
+    import argparse
+
+    parser = argparse.ArgumentParser(
+        description=(
+            'Read aircraft (AMDAR/TAMDAR/ACAR/etc.) BUFR file and convert into IODA output file')
+    )
+
+    required = parser.add_argument_group(title='required arguments')
+    required.add_argument('-i', '--input-files', nargs='+', dest='file_names',
+                          action='store', default=None, required=True,
+                          help='input files')
+    required.add_argument('-o', '--output-file', dest='output_file',
+                          action='store', default=None, required=True,
+                          help='output file')
+
+    optional = parser.add_argument_group(title='optional arguments')
+    optional.add_argument('--debug', action=argparse.BooleanOptionalAction,
+                          help='enable debug messages')
+    optional.add_argument('--verbose', action=argparse.BooleanOptionalAction,
+                          help='enable verbose debug messages')
+
+    args = parser.parse_args()
+
+    debug = True if args.debug else False
```

Comment:
If you set default values for debug and verbose, then you can simply assign the args.* values without any if statements.

Reply:
@climbfuji  Not sure I understand except to say that what's the difference here?  So I set False as one line, then I set True if args.debug is true?  Seems worse than just these simple one-liners to me.  I am still very new to Python, so I might be mis-interpreting your thought here.

Meanwhile, I have changed a bunch of my print statements to utilize the logging module - great piece of advice, thanks.

Reply:
Here is one way (not the only way) to implement this, with some additional format specifiers for the log output:
```
#!/usr/bin/env python3

import argparse
import logging

parser = argparse.ArgumentParser()
parser.add_argument('-v', '--verbose', action='count', default=0)
args = parser.parse_args()

levels = [logging.WARNING, logging.INFO, logging.DEBUG]
level = levels[min(len(levels)-1,args.verbose)]  # capped to number of levels

logging.basicConfig(level=level,
                    format="%(asctime)s %(levelname)s %(message)s")

logging.debug("a debug message")
logging.info("a info message")
logging.warning("a warning message")
```
This gives:
```
JCSDA-L-18146:Desktop heinzell$ mate logging_test.py
JCSDA-L-18146:Desktop heinzell$ python logging_test.py
2022-02-28 17:16:38,941 WARNING a warning message
JCSDA-L-18146:Desktop heinzell$ python logging_test.py -v
2022-02-28 17:16:42,460 INFO a info message
2022-02-28 17:16:42,460 WARNING a warning message
JCSDA-L-18146:Desktop heinzell$ python logging_test.py -vv
2022-02-28 17:16:44,778 DEBUG a debug message
2022-02-28 17:16:44,778 INFO a info message
2022-02-28 17:16:44,778 WARNING a warning message
```

Another way would be to stick with your definitions and use:
```
#!/usr/bin/env python3

import logging
import argparse

parser = argparse.ArgumentParser(
     description=(
         'Read aircraft (AMDAR/TAMDAR/ACAR/etc.) BUFR file and convert into IODA output file'))

optional = parser.add_argument_group(title='optional arguments')
optional.add_argument('--debug', action=argparse.BooleanOptionalAction,
                      help='enable debug messages', default = False)
optional.add_argument('--verbose', action=argparse.BooleanOptionalAction,
                      help='enable verbose debug messages', default = False)

args = parser.parse_args()

if args.debug:
    logging.basicConfig(level=logging.DEBUG)
elif args.verbose:
    logging.basicConfig(level=logging.INFO)
# This is the default
#else:
#    logging.basicConfig(level=logging.WARNING)

logging.debug("a debug message")
logging.info("a info message")
logging.warning("a warning message")
```
which gives
```
JCSDA-L-18146:Desktop heinzell$ python3 logging_test2.py
WARNING:root:a warning message
JCSDA-L-18146:Desktop heinzell$ python3 logging_test2.py --verbose
INFO:root:a info message
WARNING:root:a warning message
JCSDA-L-18146:Desktop heinzell$ python3 logging_test2.py --verbose --debug
DEBUG:root:a debug message
INFO:root:a info message
WARNING:root:a warning message
```